src/Workspaces/Core/Portable/Workspace/Workspace_Events.cs

```diff
@@ -2,291 +2,188 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
-using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Host;
-using Microsoft.CodeAnalysis.Internal.Log;
 using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.WorkspaceEventMap;
 
 namespace Microsoft.CodeAnalysis;
 
 public abstract partial class Workspace
 {
-    private readonly EventMap _eventMap = new();
+    private readonly WorkspaceEventMap _eventMap = new();
+
+    internal enum WorkspaceEventType
+    {
+        DocumentActiveContextChanged,
+        DocumentClosed,
+        DocumentOpened,
+        TextDocumentClosed,
+        TextDocumentOpened,
+        WorkspaceChange,
+        WorkspaceChangedImmediate,
+        WorkspaceFailed,
+    }
 
-    private const string WorkspaceChangeEventName = "WorkspaceChanged";
-    private const string WorkspaceChangedImmediateEventName = "WorkspaceChangedImmediate";
-    private const string WorkspaceFailedEventName = "WorkspaceFailed";
-    private const string DocumentOpenedEventName = "DocumentOpened";
-    private const string DocumentClosedEventName = "DocumentClosed";
-    private const string DocumentActiveContextChangedName = "DocumentActiveContextChanged";
-    private const string TextDocumentOpenedEventName = "TextDocumentOpened";
-    private const string TextDocumentClosedEventName = "TextDocumentClosed";
+    private IWorkspaceEventListenerService? _workspaceEventListenerService;
 
-    private IWorkspaceEventListenerService _workspaceEventListenerService;
+    #region Event Registration
 
     /// <summary>
-    /// An event raised whenever the current solution is changed.
+    /// Registers a handler that is fired whenever the current solution is changed.
     /// </summary>
-    public event EventHandler<WorkspaceChangeEventArgs> WorkspaceChanged
-    {
-        add
-        {
-            _eventMap.AddEventHandler(WorkspaceChangeEventName, value);
-        }
+    internal WorkspaceEventRegistration RegisterWorkspaceChangedHandler(Action<WorkspaceChangeEventArgs> handler, WorkspaceEventOptions? options = null)
+        => RegisterHandler(WorkspaceEventType.WorkspaceChange, handler, options);
 
-        remove
-        {
-            _eventMap.RemoveEventHandler(WorkspaceChangeEventName, value);
-        }
-    }
+    /// <summary>
+    /// Registers a handler that is fired whenever the current solution is changed.
+    /// </summary>
+    internal WorkspaceEventRegistration RegisterWorkspaceChangedImmediateHandler(Action<WorkspaceChangeEventArgs> handler, WorkspaceEventOptions? options = null)
```

Comment:
I imagine the options passed here don't matter right, since this one gets called synchronously and bypasses the queue? Should this not take options here?

Reply:
I think it's better to take in the options, for both consistency and for possible future considerations. However, I will update the comment to better reflect that the passed in threading preferences won't be adhered to.